File: integration/go/escrow_test.go
//go:build integration

package integration

import (
	"encoding/hex"
	"testing"

	"runar-integration/helpers"

	runar "github.com/icellan/runar/packages/runar-go"
)

func deployEscrow(t *testing.T, buyer, seller, arbiter *helpers.Wallet, funder *helpers.Wallet) *runar.RunarContract {
	t.Helper()

	artifact, err := helpers.CompileToSDKArtifact(
		"examples/ts/escrow/Escrow.runar.ts",
		map[string]interface{}{},
	)
	if err != nil {
		t.Fatalf("compile escrow: %v", err)
	}

	contract := runar.NewRunarContract(artifact, []interface{}{
		buyer.PubKeyHex(), seller.PubKeyHex(), arbiter.PubKeyHex(),
	})

	helpers.RPCCall("importaddress", funder.Address, "", false)
	_, err = helpers.FundWallet(funder, 1.0)
	if err != nil {
		t.Fatalf("fund: %v", err)
	}

	provider := helpers.NewRPCProvider()
	signer, err := helpers.SDKSignerFromWallet(funder)
	if err != nil {
		t.Fatalf("signer: %v", err)
	}

	_, _, err = contract.Deploy(provider, signer, runar.DeployOptions{Satoshis: 5000})
	if err != nil {
		t.Fatalf("deploy: %v", err)
	}

	return contract
}

// spendEscrowDualSig builds an unlocking script with two signatures + method index.
// The new escrow contract uses dual-signature methods:
//   - release(sellerSig, arbiterSig) — method index 0
//   - refund(buyerSig, arbiterSig) — method index 1
func spendEscrowDualSig(t *testing.T, contract *runar.RunarContract, signer1, signer2 *helpers.Wallet, methodIdx int) string {
	t.Helper()
	utxo := helpers.SDKUtxoToHelper(contract.GetCurrentUtxo())
	receiverScript := signer1.P2PKHScript()
	spendTx, err := helpers.BuildSpendTx(utxo, receiverScript, 4500)
	if err != nil {
		t.Fatalf("build spend: %v", err)
	}

	sig1Hex, err := helpers.SignInput(spendTx, 0, signer1.PrivKey)
	if err != nil {
		t.Fatalf("sign1: %v", err)
	}
	sig1Bytes, _ := hex.DecodeString(sig1Hex)

	sig2Hex, err := helpers.SignInput(spendTx, 0, signer2.PrivKey)
	if err != nil {
		t.Fatalf("sign2: %v", err)
	}
	sig2Bytes, _ := hex.DecodeString(sig2Hex)

	// Unlocking script: <sig1> <sig2> <methodIndex>
	unlockHex := helpers.EncodePushBytes(sig1Bytes) +
		helpers.EncodePushBytes(sig2Bytes) +
		helpers.EncodeMethodIndex(methodIdx)

	spendHex, err := helpers.SpendContract(utxo, unlockHex, receiverScript, 4500)
	if err != nil {
		t.Fatalf("spend: %v", err)
	}
	return spendHex
}

func TestEscrow_Compile(t *testing.T) {
	artifact, err := helpers.CompileToSDKArtifact(
		"examples/ts/escrow/Escrow.runar.ts",
		map[string]interface{}{},
	)
	if err != nil {
		t.Fatalf("compile: %v", err)
	}
	if artifact.ContractName != "Escrow" {
		t.Fatalf("expected contract name Escrow, got %s", artifact.ContractName)
	}
	t.Logf("Escrow compiled: %d bytes", len(artifact.Script)/2)
}

func TestEscrow_DeployThreePubKeys(t *testing.T) {
	buyer := helpers.NewWallet()
	seller := helpers.NewWallet()
	arbiter := helpers.NewWallet()

	artifact, err := helpers.CompileToSDKArtifact(
		"examples/ts/escrow/Escrow.runar.ts",
		map[string]interface{}{},
	)
	if err != nil {
		t.Fatalf("compile: %v", err)
	}

	contract := runar.NewRunarContract(artifact, []interface{}{
		buyer.PubKeyHex(), seller.PubKeyHex(), arbiter.PubKeyHex(),
	})

	funder := helpers.NewWallet()
	helpers.RPCCall("importaddress", funder.Address, "", false)
	_, err = helpers.FundWallet(funder, 1.0)
	if err != nil {
		t.Fatalf("fund: %v", err)
	}

	provider := helpers.NewRPCProvider()
	signer, err := helpers.SDKSignerFromWallet(funder)
	if err != nil {
		t.Fatalf("signer: %v", err)
	}

	txid, _, err := contract.Deploy(provider, signer, runar.DeployOptions{Satoshis: 5000})
	if err != nil {
		t.Fatalf("deploy: %v", err)
	}
	if len(txid) != 64 {
		t.Fatalf("expected 64-char txid, got %d", len(txid))
	}
	t.Logf("deployed with 3 distinct pubkeys: %s", txid)
}

func TestEscrow_DeploySameKey(t *testing.T) {
	wallet := helpers.NewWallet()

	artifact, err := helpers.CompileToSDKArtifact(
		"examples/ts/escrow/Escrow.runar.ts",
		map[string]interface{}{},
	)
	if err != nil {
		t.Fatalf("compile: %v", err)
	}

	contract := runar.NewRunarContract(artifact, []interface{}{
		wallet.PubKeyHex(), wallet.PubKeyHex(), wallet.PubKeyHex(),
	})

	funder := helpers.NewWallet()
	helpers.RPCCall("importaddress", funder.Address, "", false)
	_, err = helpers.FundWallet(funder, 1.0)
	if err != nil {
		t.Fatalf("fund: %v", err)
	}

	provider := helpers.NewRPCProvider()
	signer, err := helpers.SDKSignerFromWallet(funder)
	if err != nil {
		t.Fatalf("signer: %v", err)
	}

	txid, _, err := contract.Deploy(provider, signer, runar.DeployOptions{Satoshis: 5000})
	if err != nil {
		t.Fatalf("deploy: %v", err)
	}
	if len(txid) != 64 {
		t.Fatalf("expected 64-char txid, got %d", len(txid))
	}
	t.Logf("deployed with same key for all roles: %s", txid)
}

// release(sellerSig, arbiterSig) — method index 0
func TestEscrow_Release(t *testing.T) {
	buyer, seller, arbiter := helpers.NewWallet(), helpers.NewWallet(), helpers.NewWallet()
	contract := deployEscrow(t, buyer, seller, arbiter, seller)
	spendHex := spendEscrowDualSig(t, contract, seller, arbiter, 0)
	txid := helpers.AssertTxAccepted(t, spendHex)
	helpers.AssertTxInBlock(t, txid)
}

// refund(buyerSig, arbiterSig) — method index 1
func TestEscrow_Refund(t *testing.T) {
	buyer, seller, arbiter := helpers.NewWallet(), helpers.NewWallet(), helpers.NewWallet()
	contract := deployEscrow(t, buyer, seller, arbiter, buyer)
	spendHex := spendEscrowDualSig(t, contract, buyer, arbiter, 1)
	txid := helpers.AssertTxAccepted(t, spendHex)
	helpers.AssertTxInBlock(t, txid)
}

// release with wrong signer — should fail checkSig
func TestEscrow_WrongSigner_Rejected(t *testing.T) {
	buyer, seller, arbiter := helpers.NewWallet(), helpers.NewWallet(), helpers.NewWallet()
	contract := deployEscrow(t, buyer, seller, arbiter, seller)
	// Use buyer's sig where seller's is expected — should fail
	spendHex := spendEscrowDualSig(t, contract, buyer, arbiter, 0)
	helpers.AssertTxRejected(t, spendHex)
}

func TestEscrow_InvalidMethodIndex_Rejected(t *testing.T) {
	buyer, seller, arbiter := helpers.NewWallet(), helpers.NewWallet(), helpers.NewWallet()
	contract := deployEscrow(t, buyer, seller, arbiter, seller)
	// Method index 5 doesn't exist — only 0 (release) and 1 (refund)
	spendHex := spendEscrowDualSig(t, contract, seller, arbiter, 5)
	helpers.AssertTxRejected(t, spendHex)
}

File: integration/zig/src/escrow_test.zig
const std = @import("std");
const runar = @import("runar");
const bsvz = @import("bsvz");
const helpers = @import("helpers.zig");
const compile = @import("compile.zig");

test "Escrow_Compile" {
    const allocator = std.testing.allocator;

    var artifact = compile.compileContract(allocator, "examples/ts/escrow/Escrow.runar.ts") catch |err| {
        std.log.warn("Could not compile Escrow contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    try std.testing.expectEqualStrings("Escrow", artifact.contract_name);
    std.log.info("Escrow compiled: {d} bytes", .{artifact.script.len / 2});
}

test "Escrow_DeployThreePubKeys" {
    const allocator = std.testing.allocator;

    if (!helpers.isNodeAvailable(allocator)) {
        std.log.warn("Regtest node not available, skipping test", .{});
        return;
    }

    var artifact = compile.compileContract(allocator, "examples/ts/escrow/Escrow.runar.ts") catch |err| {
        std.log.warn("Could not compile Escrow contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    var buyer = try helpers.newWallet(allocator);
    defer buyer.deinit();
    var seller = try helpers.newWallet(allocator);
    defer seller.deinit();
    var arbiter = try helpers.newWallet(allocator);
    defer arbiter.deinit();

    const buyer_pk = try buyer.pubKeyHex(allocator);
    defer allocator.free(buyer_pk);
    const seller_pk = try seller.pubKeyHex(allocator);
    defer allocator.free(seller_pk);
    const arbiter_pk = try arbiter.pubKeyHex(allocator);
    defer allocator.free(arbiter_pk);

    var contract = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = buyer_pk },
        .{ .bytes = seller_pk },
        .{ .bytes = arbiter_pk },
    });
    defer contract.deinit();

    // Fund the seller (who will pay for deployment)
    var funder = try helpers.newWallet(allocator);
    defer funder.deinit();
    const fund_txid = try helpers.fundWallet(allocator, &funder, 1.0);
    defer allocator.free(fund_txid);

    var rpc_provider = helpers.RPCProvider.init(allocator);
    var local_signer = try funder.localSigner();

    const deploy_txid = try contract.deploy(rpc_provider.provider(), local_signer.signer(), .{ .satoshis = 5000 });
    defer allocator.free(deploy_txid);

    try std.testing.expectEqual(@as(usize, 64), deploy_txid.len);
    std.log.info("Escrow deployed with 3 distinct pubkeys: {s}", .{deploy_txid});
}

test "Escrow_DeploySameKey" {
    const allocator = std.testing.allocator;

    if (!helpers.isNodeAvailable(allocator)) {
        std.log.warn("Regtest node not available, skipping test", .{});
        return;
    }

    var artifact = compile.compileContract(allocator, "examples/ts/escrow/Escrow.runar.ts") catch |err| {
        std.log.warn("Could not compile Escrow contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    // All three roles use the same key
    var wallet = try helpers.newWallet(allocator);
    defer wallet.deinit();
    const pk_hex = try wallet.pubKeyHex(allocator);
    defer allocator.free(pk_hex);

    var contract = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = pk_hex },
        .{ .bytes = pk_hex },
        .{ .bytes = pk_hex },
    });
    defer contract.deinit();

    var funder = try helpers.newWallet(allocator);
    defer funder.deinit();
    const fund_txid = try helpers.fundWallet(allocator, &funder, 1.0);
    defer allocator.free(fund_txid);

    var rpc_provider = helpers.RPCProvider.init(allocator);
    var local_signer = try funder.localSigner();

    const deploy_txid = try contract.deploy(rpc_provider.provider(), local_signer.signer(), .{ .satoshis = 5000 });
    defer allocator.free(deploy_txid);

    try std.testing.expectEqual(@as(usize, 64), deploy_txid.len);
    std.log.info("Escrow deployed with same key for all roles: {s}", .{deploy_txid});
}

test "Escrow_ABI_Methods" {
    const allocator = std.testing.allocator;

    var artifact = compile.compileContract(allocator, "examples/ts/escrow/Escrow.runar.ts") catch |err| {
        std.log.warn("Could not compile Escrow contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    // Escrow should have at least 2 public methods: release and refund
    var public_count: usize = 0;
    var has_release = false;
    var has_refund = false;
    for (artifact.abi.methods) |m| {
        if (m.is_public) {
            public_count += 1;
            if (std.mem.eql(u8, m.name, "release")) has_release = true;
            if (std.mem.eql(u8, m.name, "refund")) has_refund = true;
        }
    }
    try std.testing.expect(public_count >= 2);
    try std.testing.expect(has_release);
    try std.testing.expect(has_refund);
}

test "Escrow_NotStateful" {
    const allocator = std.testing.allocator;

    var artifact = compile.compileContract(allocator, "examples/ts/escrow/Escrow.runar.ts") catch |err| {
        std.log.warn("Could not compile Escrow contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    // Escrow is a stateless contract
    try std.testing.expect(!artifact.isStateful());
}
